lib/astarte/utilities/params_gen.ex

```diff
@@ -0,0 +1,175 @@
+  """
+  @spec __using__(opts :: any()) :: Macro.t()
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__)
+      import StreamData
+      import ExUnitProperties
+    end
+  end
+
+  @doc false
+  @spec params({:gen, any(), [{:all, any(), any()}]}) :: Macro.t()
+  defmacro params({:gen, _gen_meta, [{:all, _all_meta, clauses_with_body}]}) do
+    {clauses, [[do: body]]} = Enum.split(clauses_with_body, -1)
+    compile(clauses, body, __CALLER__)
+  end
+
+  @doc """
+  Defines a custom property-based test generator macro (`params gen all`) that supports overriding default generators.
+  It processes generator clauses and applies any custom overrides provided via the `:params` keyword.
+  """
+  @spec params({:gen, any(), [{:all, any(), list()}]}, [{:do, any()}]) :: Macro.t()
+  defmacro params({:gen, _gen_meta, [{:all, _all_meta, clauses}]}, do: body) do
+    compile(clauses, body, __CALLER__)
+  end
+
+  @streamdata_functions StreamData.__info__(:functions) |> Enum.map(&elem(&1, 0))
+  defp stream_data?(_aliases, {func, _func_meta, _}) when func in @streamdata_functions,
+    do: true
+
+  defp stream_data?(
+         aliases,
+         {{:., _dot_meta, [{:__aliases__, _aliases_meta, alias_list}, func]}, _other_meta, _rest}
+       )
+       when func in @streamdata_functions do
+    Module.concat(alias_list) in aliases
+  end
+
+  defp stream_data?(_, _), do: false
+
+  defp override(aliases, {:<-, meta, [var_and_meta, _]}, value) do
+    command =
+      if stream_data?(aliases, value) do
+        quote do: unquote(value)
+      else
+        quote do
+          constant(unquote(value))
+        end
+      end
+
+    {:<-, meta, [var_and_meta, command]}
+  end
+
+  defp compile_aliases([], acc), do: [StreamData | acc]
+
+  defp compile_aliases([{alias, StreamData} | tail], acc) do
+    compile_aliases(tail, [alias | acc])
+  end
+
+  defp compile_aliases(_, []), do: []
+  defp compile_aliases(aliases), do: compile_aliases(aliases, [])
+
+  defp compile_clauses(_aliases, [], _, acc), do: acc
+
+  defp compile_clauses(aliases, [{:<-, _, [{var, _, _}, _]} = clause | tail], params, acc) do
+    clause =
+      if Keyword.has_key?(params, var) do
+        override(aliases, clause, params[var])
+      else
+        clause
+      end
+
+    compile_clauses(aliases, tail, params, [clause | acc])
+  end
+
+  defp compile_clauses(aliases, clauses, params),
+    do: compile_clauses(aliases, clauses, params, [])
+
+  defp compile(clauses, body, %{aliases: aliases}) do
+    {params, clauses} =
+      case List.last(clauses) do
+        [params: params] when is_list(params) ->
+          {params, List.delete_at(clauses, length(clauses) - 1)}
+
+        _ ->
+          {[], clauses}
+      end
+
+    clauses =
+      aliases
+      |> compile_aliases()
+      |> compile_clauses(clauses, params)
+      |> Enum.reverse()
```

Comment:
then we can avoid checking the aliases and stuff. This was error prone anyway tbh

```suggestion
    clauses = compile_clauses(clauses, params, env)
```

`env` here is the (whole) third argument of your function